### PR TITLE
Add proposal for container recreate request

### PR DESCRIPTION
**Pull request by FillZpp**
Signed-off-by: FillZpp <FillZpp.pub@gmail.com>

### Ⅰ. Describe what this PR does
Add proposal for container recreate request.

Markdown link: https://github.com/FillZpp/kruise/blob/proposal-for-container-recreate-request/docs/proposals/20210316-containerrecreaterequest.md

### Ⅱ. Does this pull request fix one issue?

related #495

### CODE REVIEW

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+Now that the only way to recreate a container in an active Pod is to update it in-place.
+However, it also requires users to change the image defined in workloads (e.g. CloneSet/Advanced StatefulSet).
+
+We hope to provide a way to let users recreate(restart) one or more containers in any active Pods.
+It is useful for some user scenarios, such as:
+
+1. Restart the application in app container.
+2. Restart a sidecar contaienr along with app container.
+3. Force the init container running once again.
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:41]: please identify the side-effect of rerun the init container, and maybe kruise should not cover such feature for now

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    crr.apps.kruise.io/pod-name: pod-name
+    crr.apps.kruise.io/node-name: node-name-of-the-pod  # injected by kruise webhook
+spec:
+  initContainers:
+  - name: app-init
+  containers:
+  - name: app     # if no preStop specified, we will execute the preStop hook defined in Pod spec
+  - name: sidecar
+    lifecycle:
+      preStop:    # this will overwrite the preStop hook defined in Pod spec
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:63]: please provide user cases about why the prestop hook should be overwritten

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  ttlSecondsAfterFinished: 1800     # ttl after this ContainerRecreateRequest has finished
+```
+
+1. The initContainers are guaranteed to re-run before containers to start recreating.
+2. The containers will be recreated one by one.
+   Typically, we will stop the next one once if the previous has stopped, unless the `forwardByReady` is `true`.
+3. The `readyProtection` feature depends on a new feature-gate named `PodRecreateReadinessGate`,
+   which indicates to inject a readinessGate during each Pod creating.
+
+### Implementation Details/Notes/Constraints
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:85]: a separate user cases should be provided

**Review by furykerry — CHANGES_REQUESTED** [review]

File: docs/proposals/20210316-containerrecreaterequest.md (hunk 1 above)
Comment (FillZpp) [reply docs/proposals/20210316-containerrecreaterequest.md]: removed temporarily

File: docs/proposals/20210316-containerrecreaterequest.md (hunk 2 above)
Comment (FillZpp) [reply docs/proposals/20210316-containerrecreaterequest.md]: removed temporarily

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+spec:
+  containers:
+  - name: app
+  - name: sidecar
+  strategy:
+    skipPreStop: false      # 'true' means do not execute preStop hook before stop the container
+    gracePeriodSeconds: 30  # the timeout period for preStop hook and container stopping
+    failurePolicy: Fail     # or 'Ignore'
+    readyProtection: false  # 'true' means make the Pod not-ready before recreating containers
+    forwardByReady: false   # 'true' means recreate the next container only if the previous one has been recreated and ready
```
Comment (Fei-Guo) [inline docs/proposals/20210316-containerrecreaterequest.md:66]: forwardByReady-> continueByReady?

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+However, it also requires users to change the image defined in workloads (e.g. CloneSet/Advanced StatefulSet).
+
+We hope to provide a way to let users recreate(restart) one or more containers in any active Pods.
+It is useful for some user scenarios, such as:
+
+1. Restart the application in app container.
+2. Restart a sidecar contaienr along with app container.
+
+## Proposal
+
```
Comment (Fei-Guo) [inline docs/proposals/20210316-containerrecreaterequest.md:45]: Add a prerequisite section, 

Kruise-daemon is required. Btw, we should a have a document to describe kruise-daemon. IMO, it is a break through and allows us to do a lot of advanced maintenance work through K8s controllers.

**Review by Fei-Guo — COMMENTED** [review]
Overall LG.

**Comment by Fei-Guo** [comment]
LGTM

**Comment by tudou34** [comment]
can add a strategy like recreat-strategy in deployment? it can recreate all containers  at one time. @FillZpp

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  - name: app
+  - name: sidecar
+    skipPreStop: false           # 'true' means do not execute preStop hook before stop the container
+  strategy:
+    failurePolicy: Fail          # or 'Ignore'
+    stopGracePeriodSeconds: 30   # the timeout period for preStop hook and container stopping
+    readyProtection: false       # 'true' means make the Pod not-ready before recreating containers
+    continueByReady: false       # 'true' means recreate the next container only if the previous one has been recreated and ready
+  activeDeadlineSeconds: 300        # deadline of this ContainerRecreateRequest
+  ttlSecondsAfterFinished: 1800     # ttl after this ContainerRecreateRequest has finished
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:68]: please add ContainerRecreateRequest status description, and describe how user determine the recreate is successful

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  labels:
+    crr.apps.kruise.io/pod-name: pod-name
+    crr.apps.kruise.io/node-name: node-name-of-the-pod  # injected by kruise webhook
+spec:
+  containers:
+  - name: app
+  - name: sidecar
+    skipPreStop: false           # 'true' means do not execute preStop hook before stop the container
+  strategy:
+    failurePolicy: Fail          # or 'Ignore'
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:63]: please comments about the meaning of failurePolicy

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+![container_recreate](../../docs/img/containerrecreate.png)
+
+For containers, **kruise-daemon** will stop the container with or without preStop hook,
+then **kubelet** will create a new container and start again.
+
+While some containers being recreating, the other containers in the same Pod will still be running.
+
+### Risks and Mitigations
+
+- Recreating a container will cause the application in it to start again.
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:102]: is it possible that the recreation and inplace-update of one pod occur in the same time? please describe the possible problem and solutions.

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    continueByReady: false       # 'true' means recreate the next container only if the previous one has been recreated and ready
+  activeDeadlineSeconds: 300        # deadline of this ContainerRecreateRequest
+  ttlSecondsAfterFinished: 1800     # ttl after this ContainerRecreateRequest has finished
+```
+
+1. The containers will be recreated one by one.
+   Typically, we will stop the next one once if the previous has stopped, unless the `continueByReady` is `true`.
+2. The `readyProtection` feature depends on a new feature-gate named `PodRecreateReadinessGate`,
+   which indicates to inject a readinessGate during each Pod creating.
+
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:75]: please describe how to recycle the finished ContainerRecreateRequest cr

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+  containers:
+  - name: app
+  - name: sidecar
+    skipPreStop: false           # 'true' means do not execute preStop hook before stop the container
+  strategy:
+    failurePolicy: Fail          # or 'Ignore'. If 'Fail', the CRR will finish when one container failed to stop or recreate
+    stopGracePeriodSeconds: 30   # the timeout period for preStop hook and container stopping
+    readyProtection: false       # 'true' means make the Pod not-ready before recreating containers
+    continueByReady: false       # 'true' means recreate the next container only if the previous one has been recreated and ready
+  activeDeadlineSeconds: 300        # deadline of this ContainerRecreateRequest
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:67]: what does deadline means here ? what will happen after the deadline?

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    crr.apps.kruise.io/pod-name: pod-name
+    crr.apps.kruise.io/node-name: node-name-of-the-pod  # injected by kruise webhook
+spec:
+  containers:
+  - name: app
+  - name: sidecar
+    skipPreStop: false           # 'true' means do not execute preStop hook before stop the container
+  strategy:
+    failurePolicy: Fail          # or 'Ignore'. If 'Fail', the CRR will finish when one container failed to stop or recreate
+    stopGracePeriodSeconds: 30   # the timeout period for preStop hook and container stopping
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:64]: can we implement skipprestop=true by set stopGracePeriodSeconds to zero ? and is it possible to set different stopGracePeriodSeconds for different containers in a single pod

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    readyProtection: false       # 'true' means make the Pod not-ready before recreating containers
+    continueByReady: false       # 'true' means recreate the next container only if the previous one has been recreated and ready
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:66]: consider rename continueByReady to orderedRecreate

File: docs/proposals/20210316-containerrecreaterequest.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    crr.apps.kruise.io/node-name: node-name-of-the-pod  # injected by kruise webhook
+spec:
+  containers:
+  - name: app
+  - name: sidecar
+    skipPreStop: false           # 'true' means do not execute preStop hook before stop the container
+  strategy:
+    failurePolicy: Fail          # or 'Ignore'. If 'Fail', the CRR will finish when one container failed to stop or recreate
+    stopGracePeriodSeconds: 30   # the timeout period for preStop hook and container stopping
+    readyProtection: false       # 'true' means make the Pod not-ready before recreating containers
```
Comment (furykerry) [inline docs/proposals/20210316-containerrecreaterequest.md:65]: how long must  we wait before recreating a container ?

**Review by furykerry — CHANGES_REQUESTED** [review]

File: docs/proposals/20210316-containerrecreaterequest.md (hunk 10 above)
Comment (FillZpp) [reply docs/proposals/20210316-containerrecreaterequest.md]: done

File: docs/proposals/20210316-containerrecreaterequest.md (hunk 11 above)
Comment (FillZpp) [reply docs/proposals/20210316-containerrecreaterequest.md]: done

File: docs/proposals/20210316-containerrecreaterequest.md (hunk 12 above)
Comment (FillZpp) [reply docs/proposals/20210316-containerrecreaterequest.md]: done

File: docs/proposals/20210316-containerrecreaterequest.md (hunk 13 above)
Comment (FillZpp) [reply docs/proposals/20210316-containerrecreaterequest.md]: have replaced `readyProtection` to `unreadyGracePeriodSeconds `

**Review by furykerry — APPROVED** [review]

**Pull request merged**